### PR TITLE
[FLINK-10397] Remove CoreOptions#MODE

### PR DESCRIPTION
**Pull request by tillrohrmann**
## What is the purpose of the change

This PR is based on #6748 and #6751 and is part of the [removal of Flink's legacy mode](https://issues.apache.org/jira/browse/FLINK-10392).

Removes the CoreOptions#MODE option used to switch between the new and legacy mode.

## Brief change log

- Remove `CoreOptions#MODE`, `CoreOptions#NEW` and `CoreOptions#LEGACY`
- Set the mode where it was switchable to new (e.g. `isNewMode = true` in `YarnTestBase`)

## Verifying this change

- Not explicitly tested

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by yanghua** [comment]
@tillrohrmann The travis reports two issues : 

```
  JobManagerHAProcessFailureBatchRecoveryITCase.testDispatcherProcessFailure:331 The program did not finish in time
```
and 

```
Running org.apache.flink.docs.configuration.ConfigOptionsDocsCompletenessITCase
Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.771 sec <<< FAILURE! - in org.apache.flink.docs.configuration.ConfigOptionsDocsCompletenessITCase
testFullReferenceCompleteness(org.apache.flink.docs.configuration.ConfigOptionsDocsCompletenessITCase)  Time elapsed: 0.724 sec  <<< FAILURE!
java.lang.AssertionError: Documentation is outdated, please regenerate it according to the instructions in flink-docs/README.md.
	Problems:
		Documented option mode does not exist.
	at org.junit.Assert.fail(Assert.java:88)
	at org.apache.flink.docs.configuration.ConfigOptionsDocsCompletenessITCase.compareDocumentedAndExistingOptions(ConfigOptionsDocsCompletenessITCase.java:118)
	at org.apache.flink.docs.configuration.ConfigOptionsDocsCompletenessITCase.testFullReferenceCompleteness(ConfigOptionsDocsCompletenessITCase.java:76)


Results :

Failed tests: 
  ConfigOptionsDocsCompletenessITCase.testFullReferenceCompleteness:76->compareDocumentedAndExistingOptions:118 Documentation is outdated, please regenerate it according to the instructions in flink-docs/README.md.
	Problems:
		Documented option mode does not exist.
```

**Comment by TisonKun** [comment]
@yanghua The former can be discussed at FLINK-10403 and the latter IMO can be resolved by regenerate the doc( not a critical blocker ).

**Comment by yanghua** [comment]
@TisonKun  I know, I just give a reminder so that Travis can be turned green as soon as possible.

File: flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java [hunk 1]
```diff
@@ -308,22 +308,4 @@
 	// ------------------------------------------------------------------------
 	//  Distributed architecture
```
Comment (StefanRRichter) [inline flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java]: Comment section can be removed.

File: flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendTestBase.java [hunk 2]
```diff
@@ -19,43 +19,21 @@
...
 import org.apache.flink.util.TestLogger;
 
-import org.junit.runner.RunWith;
-import org.junit.runners.Parameterized;
-
-import java.util.Arrays;
-import java.util.List;
-
 /**
  * Base test class for {@link CliFrontend} tests that wraps the new vs. legacy mode.
```
Comment (StefanRRichter) [inline flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendTestBase.java]: Comment update required?

**Comment by TisonKun** [comment]
@StefanRRichter I would say that the removal of `EXECUTION_MODE_OPTION` and `isNewMode` will be part of [FLINK-10392](https://issues.apache.org/jira/browse/FLINK-10392). For the `isNewMode` part, see also my comment under #6748 .

File: flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendTestBase.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendTestBase.java]: Good catch. Will update it.

**Comment by tillrohrmann** [comment]
I agree with @TisonKun. Removing the `EXECUTION_MODE_OPTION` and `isNewMode` can be done in another subtask of FLINK-10392.

**Pull request merged**